### PR TITLE
Implement and action for manually clearing the Terminal (and conpty) buffer

### LINKED ISSUE #1193: Support manually clearing the conpty buffer
*From @Tyriar in [microsoft/vscode#75141](https://github.com/microsoft/vscode/issues/75141)*

> This is happening because Terminal: Clear tells the frontend (the thing rendering everything) to clear, but the way that the conpty backend works is it will sometimes reprint what it thinks should be visible. Since it doesn't know about our clear it reprints the unwanted output, this doesn't happen on a regular linux/mac backend that conpty is trying to emulate.

> Is there anything that can be done here to tell conpty to not reprint the screen? I suspect python.exe does not enter alt mode so I'm not sure why a reprint is necessary.

### LINKED ISSUE #1882: Add a terminal-side shortcut for clearing the screen and/or scrollback
It would be great to have the equivalent of Command+K on the Mac terminal.  This clears the terminal completely, not the equivalent of typing "clear", which only scrolls the cursor to the top of the terminal.  My use case is the following:
- Open an Ubuntu terminal
- ssh into a server
- run a command which may produce large amounts of content and exceptions within the content
- fix the exceptions by making a code change
- CLEAR the terminal completely (not just a "clear" command)
- re-run the command
- now that the previous command output was cleared, we can search for new exceptions without getting the previous exceptions
- NOTE: this also requires a "ctrl+F" type feature to exist in the terminal, which currently does not

### PR DESCRIPTION
**Pull request by zadjii-msft**
## Summary of the Pull Request

clear-buffer-000

This adds a new action, `clearBuffer`. It accepts 3 values for the `clear` type:
* `"clear": "screen"`: Clear the terminal viewport content. Leaves the scrollback untouched. Moves the cursor row to the top of the viewport (unmodified).
* `"clear": "scrollback"`: Clear the scrollback. Leaves the viewport untouched.
* `"clear": "all"`: (**default**) Clear the scrollback and the visible viewport. Moves the cursor row to the top of the viewport (unmodified).

"Clear Buffer" has also been added to `defaults.json`.

## References
* From microsoft/vscode#75141 originally

## PR Checklist
* [x] Closes #1193
* [x] Closes #1882
* [x] I work here
* [x] Tests added/passed
* [ ] Requires documentation to be updated

## Detailed Description of the Pull Request / Additional comments

This is a bit tricky, because we need to plumb it all the way through conpty to clear the buffer. If we don't, then conpty will immediately just redraw the screen. So this sends a signal to the attached conpty, and then waits for conpty to draw the updated, cleared, screen back to us.

## Validation Steps Performed
* works for each of the three clear types as expected
* tests pass.
* works even with `ping -t 8.8.8.8` as you'd hope.

### CODE REVIEW

**Review by xdaolai — APPROVED** [review]

File: src/host/PtySignalInputThread.hpp [hunk 1]
```diff
@@ -35,6 +35,7 @@ namespace Microsoft::Console
     private:
         enum class PtySignal : unsigned short
         {
+            ClearBuffer = 2,
```
[the remark this answers is not in the document]
Comment (zadjii-msft) [reply src/host/PtySignalInputThread.hpp:38]: Honestly, no good reason. These signal values are totally arbitrary. When I think of "clear screen", I think of `^[[2J`, so I went with that. Kinda like "resize buffer" is reminiscent of `^[[8;<r>;<c>t`

File: src/host/screenInfo.cpp [hunk 2]
```diff
@@ -2277,6 +2277,55 @@ void SCREEN_INFORMATION::SetViewport(const Viewport& newViewport,
...
+// - <none>
+// Return Value:
+// - S_OK
+[[nodiscard]] HRESULT SCREEN_INFORMATION::ClearBuffer()
+{
+    const COORD oldCursorPos = _textBuffer->GetCursor().GetPosition();
+    short sNewTop = oldCursorPos.Y;
+    const Viewport oldViewport = _viewport;
+
+    short delta = (sNewTop + _viewport.Height()) - (GetBufferSize().Height());
```
Comment (carlos-zamora) [inline src/host/screenInfo.cpp:2296]:
```suggestion
    const short delta = (sNewTop + _viewport.Height()) - (GetBufferSize().Height());
```

File: src/host/screenInfo.cpp [hunk 3]
```diff
@@ -2277,6 +2277,55 @@ void SCREEN_INFORMATION::SetViewport(const Viewport& newViewport,
...
+    // Update all the rows in the current viewport with the standard erase attributes,
+    // i.e. the current background color, but with no meta attributes set.
+    auto fillAttributes = GetAttributes();
+    fillAttributes.SetStandardErase();
+
+    // +1 on the y coord because we don't want to clear the attributes of the
+    // cursor row, the one we saved.
+    auto fillPosition = COORD{ 0, _viewport.Top() + 1 };
+    auto fillLength = gsl::narrow_cast<size_t>(_viewport.Height() * GetBufferSize().Width());
+    auto fillData = OutputCellIterator{ fillAttributes, fillLength };
```
Comment (carlos-zamora) [inline src/host/screenInfo.cpp:2318]:
```suggestion
    const auto fillPosition = COORD{ 0, _viewport.Top() + 1 };
    const auto fillLength = gsl::narrow_cast<size_t>(_viewport.Height() * GetBufferSize().Width());
    const auto fillData = OutputCellIterator{ fillAttributes, fillLength };
```
Not sure if these can be const, but if they can, there ya go.

**Review by carlos-zamora — APPROVED** [review]
Pretty confident with the Terminal stuff (except maybe TermConnection). The rest looks good too, but I'm a lot less familiar with that area.

File: src/host/PtySignalInputThread.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (miniksa) [reply src/host/PtySignalInputThread.hpp:38]: K. I thought there was some secret I didn't know. That's logical enough.

**Review by miniksa — APPROVED** [review]
Okay, gonna approve. It is what it is.

I don't really like that we have to add another signal. Signaling to do the equivalent of SIGWNCH is well known and is what 8 is for, but this clear thing seems like a new invention which I'm not totally happy with. However... it's a problem that Windows sort of ends up with thanks to our weird architecture so I guess having our own solution is what's in order. Also it's just code so if we're proven wrong, we can just change it.

I want it to be like a VT signal on one of the existing channels, but on the input channel it makes no sense. And to get to the output channel, the alternative of making the Windows Terminal take a client-side reference to the same console as the client-app hosted in the ConPTY just to call clear seems ridiculous.

I think this is going to become less relevant over time as we approach passthrough mode where ConPTY will be less integral to the rendering on the Windows Terminal. But for now... the code must flow.

**Pull request merged**